### ISSUE TITLE
ENV vs. lane_context vs. SharedValues?

### ISSUE THREAD

**Issue by mollyIV** [fastlane/fastlane#14255]
Few days ago the issue related to `lane_context` and ENV has been created https://github.com/fastlane/fastlane/issues/14182. The investigation opened up a broader discussion about the usage of `SharedValues` in general. I talked to @janpio and we came up with a few questions / ideas and decided to open a new issue for the further discussion. 

**When should a user use the values from `lane_context[SharedValues::XYZ]` and when should a user use the values from ENV?** 

`SharedValues` and ENV seem to be very close with each other. Could you please provide a bit more context what was the reason behind introducing `SharedValues` and not just using ENV variables in order to pass the values between lanes? 

I couldn't find the docs saying about the relationship between the values available as `SharedValues` and the values available as environment variables (ENV). Currently there's no much documentation about [lane_context](https://docs.fastlane.tools/actions/lane_context/):

* https://docs.fastlane.tools/advanced/lanes/#lane-context

> The different actions can communicate with each other using a shared hash.

The documentation basically says you can use the shared hash between the lanes but there's missing info which lanes expose which values which points us to the question number two 👇 

**It is not possible to print out the ENV and `SharedValues` values which are set by a particular lane.** 

Couldn't find any info how to print out the values set by a lane. What we can do though is to print out  ENV variables (`fastlane action [action_name]`) that we can set as input parameters instead of passing input parameters directly to the lane. We could add a new functionality of printing out the summary table and it would be very helpful to see what `SharedValues` / ENV action/lane sets. In that way a user could easily spot what `SharedValues`/ENV will be set by a particular lane/action. 

**How to decide whether a user should set `SharedValues` or ENV in his plugin or new lane?**

Since there's no clear rules to follow whether to use `SharedValues` or ENV variables it's hard to decide what to support in new lanes / custom plugins. 

We'd love to hear your thoughts on this subject and come up with the ideas how we could clean up the usage of `SharedValue`, ENV and `lane_context` and make them easier to understand 🚀  

Thank you! 🙇 🙇

**Comment by janpio** [comment]
I think the question is a bit broader even than what you suggested:

> How should fastlane actions, plugins and user lanes work with and manipulate `ENV` variables, `lane_context` and `SharedValues`?

Different problem areas are:

- Usage Documentation
- Development Documentation
- Development Practices

---

My current knowledge:

### ENV variables 

- They are used to configure fastlane in general (example: https://docs.fastlane.tools/advanced/fastlane/#configuring-fastlane)
- They are used as input and output but also for communication between lanes. 
- All (?) action options have an associated environment variable that you can set to give that param a value (see https://docs.fastlane.tools/advanced/fastlane/#priorities-of-parameters-and-options and https://docs.fastlane.tools/advanced/actions/#resolution-order for priority of different methods)
- I could not find any specific documentation on how/where/why actions can/should use ENV variables to _output_ anything, but I know several do.
- As I understand it, setting ENV vars is mostly used for inter action communication to make values used/created/set in an action available to other code/actions

### lane context

- `SharedValues` is a list of "keys" that defines what is set by an action
- `lane_context` is the "tool" to get the values belonging to those keys (See https://docs.fastlane.tools/advanced/lanes/#lane-properties and https://docs.fastlane.tools/advanced/lanes/#lane-context)
- How exactly this is structure or defined, I don't know
- I remember that the lane context also provides some general information like the current lane name, maybe also the platform etc. So it could be used for "environment" information as well.

---

Sometime lane context is made available to users already: If an action crashes:

```
[2019-02-16 20:09:36.55]: Lane Context:
[2019-02-16 20:09:36.55]: {:PLATFORM_NAME=>nil, :LANE_NAME=>"build", :XCODEBUILD_DERIVED_DATA_PATH=>nil}
```

---

Here is an older issue where we looked into this a bit a few ~~months~~ years ago: https://github.com/fastlane/docs/issues/468

---

Here is a param documentation `flavor` that mentiones lane context and SharedValues: https://docs.fastlane.tools/actions/build_android_app/#parameters
This one mentions "shared values" in the description: https://docs.fastlane.tools/actions/xcode_server_get_assets/#xcode_server_get_assets

---

Googling around returns lots of 3rd party articles that mentions SharedValues and lane context:

https://books.google.de/books?id=yOlODwAAQBAJ&pg=PA302&lpg=PA302&dq=fastlane+SharedValues&source=bl&ots=ZVjpUD02yN&sig=ACfU3U2Qz4bxrvLwRytl6rYgZBfRZLatfg&hl=de&sa=X&ved=2ahUKEwjM76TXvM3gAhVLZVAKHTXUAbc4ChDoATABegQICBAB#v=onepage&q=fastlane%20SharedValues&f=false
https://blog.novoda.com/become-a-master-of-automation-with-fastlane/
https://stackoverflow.com/questions/46192561/how-do-i-get-a-shared-value-in-fastlane
https://teamairship.com/wp-content/uploads/2017/03/sign_apk.txt
https://www.eharmony.com/engineering/fastlane-ios-release-automation/
https://medium.com/@faizmokhtar/how-to-improve-your-beta-deployment-flow-with-fastlane-badge-plugin-289e2ea5d2bb
https://dbanck.svbtle.com/part-2-android-with-fastlane-supply
https://docs.deploygate.com/docs/fastlane
https://www.hedgehoglab.com/blog/app-developer-fastlane-review
https://de.slideshare.net/XebiaFrance/xebicon16-fastlane-automatisez-votre-vie-de-dveloppeur-ios-par-jeanchristophe-pastant-dveloppeur-mobile-chez-xebia
https://raw.githubusercontent.com/thedoritos/flaction/master/fastlane/actions/balto.rb
https://rolandleth.com/tech/blog/fastlane-fastfile-3
https://dev.welaika.com/blog/2016/12/01/fastlane.html
https://pixyzehn.com/2016/08/24/setup-fastlane-with-circleci.html

**Comment by janpio** [comment]
So after going through all that, I think I can summarize my understanding: 

Fastlane uses envirnment variables to accept and give information to outside of its context (What is known before fastlane is run? What should be known after it has finished?), and uses an internal `lane_context` hash to exchange data (that can be better defined with `SharedValues`) in its internal parts.

Currently both elements are only partially documented and discoverable unless you read the actual source code.

@joshdholtz You sure have some input here, did we miss anything?

**Comment by janpio** [comment]
One problem in generating documentation could be that the `SharedValues` module inside an Action module only contains the SharedValues _created_ by the action:

https://github.com/fastlane/fastlane/blob/69a77679a58868d202e36fa14317e841cda4f5ef/fastlane/lib/fastlane/actions/xcodebuild.rb#L4-L7
https://github.com/fastlane/fastlane/blob/69a77679a58868d202e36fa14317e841cda4f5ef/fastlane/lib/fastlane/actions/xcodebuild.rb#L92
https://github.com/fastlane/fastlane/blob/69a77679a58868d202e36fa14317e841cda4f5ef/fastlane/lib/fastlane/actions/xcodebuild.rb#L118
https://github.com/fastlane/fastlane/blob/69a77679a58868d202e36fa14317e841cda4f5ef/fastlane/lib/fastlane/actions/xcodebuild.rb#L130
https://github.com/fastlane/fastlane/blob/69a77679a58868d202e36fa14317e841cda4f5ef/fastlane/lib/fastlane/actions/xcodebuild.rb#L149

`IPA_OUTPUT_PATH` "belongs" to some other action (probably `gym` I guess) and so is not listed in the `SharedValues` here, but used in the source code.

**Comment by janpio** [comment]
Some more poking around:

- lane_context/SharedValues are often used as `default_value` of action options
- Sometimes env vars are used as fallbacks to lane_context: https://github.com/fastlane/fastlane/blob/69a77679a58868d202e36fa14317e841cda4f5ef/fastlane/lib/fastlane/actions/apteligent.rb#L33
- As setting a lane_context value requires using `Actions.lane_context`, we could probably create a "log" of its usage that could be output when using an action to give insight what is going on _at runtime_
- It should be pretty simple to generate a list of SharedValues and their usage (which file and method?) across the whole codebase
- Also: Who does a SharedValue "belong" to (defined in `SharedValues` module), and who uses it?
- Which one are used without belonging to anyone? What lane_context keys are used directly without going through `SharedValues` (if there are any)?

- verify blocks in action options are used to set the option value in an env var https://github.com/fastlane/fastlane/blob/69a77679a58868d202e36fa14317e841cda4f5ef/cert/lib/cert/options.rb#L48-L50
   - Why do we do this, copy an option into an ENV var? Is is not super easy to leak sensitive data that way?
- Env vars are also used deeper to configure stuff (vs. just as alternatives to options of actions): https://github.com/fastlane/fastlane/blob/69a77679a58868d202e36fa14317e841cda4f5ef/fastlane_core/lib/fastlane_core/itunes_transporter.rb#L147-L156
-

**Comment by janpio** [comment]
A good first step could be to output all the `SharedValues` modules in the documentation (CLI and online). "This action uses the following `SharedValues` to make information available via `lane_context`" or similar could be the description of the table. (SharedValues and lane_context link to an appropriate advanced documentation)

**Comment by hydraSlav** [comment]
> It is not possible to print out the ENV and `SharedValues`

Not sure about ENV, but `debug` statement after any action will print all SharedValues currently available, not just from the last action (but of course it updates after every action)

**Comment by mollyIV** [comment]
Thank you so much for your findings @janpio 🏅 🏅 

I really like the way you think of ENV and `lane_context`:
* using ENV as input parameters 
* using ENV as output parameters
* using `lane_context` as inner lanes communication 

Printing ENV as input parameters is already done and it is well documented. 
It would be great to also print ENV as the output from particular lane - summary table generated automatically. 

Speaking about the inner lanes communication I am wondering if `lane_context` values should be even exposed to client usage?  Would be possible to have private storage (in memory?) which would be only accessible by inner source code to pass the params between the lanes? 🤔
 
ENV variables would only play a role of public API -  it is just raw idea 🤔

> A good first step could be to output all the `SharedValues` modules in the documentation (CLI and online).

Absolutely 💪 You think that would be possible to do by typing a command per lane, i.e. adding `SharedValues` summary table to `fastlane action [action_name]`? If so, could you please share the technical details? 🙏

**Comment by janpio** [comment]
> It would be great to also print ENV as the output from particular lane - summary table generated automatically.

Yeah, problem here is that this can happen anywhere in the action code or the classes used by an action. During runtime we could take snapshots of ENV before and after the action, and then output a diff - but for the documentation I can't really think of a good way. (Except that we require/suggest that an action maintains a list of used ENV vars and a description of the how/why/when in the action definition).

> Speaking about the inner lanes communication I am wondering if `lane_context` values should be even exposed to client usage? Would be possible to have private storage (in memory?) which would be only accessible by inner source code to pass the params between the lanes? 🤔

That depends on how you define "internal" or "inner". From a technical standpoint, everything happening in ruby is internal - and is meant to be. There is an explicit `Action.lane_context` method to access and change these variables, and all ruby - the lanes, the actions, plugin actions, local lanes, normal ruby methods in lanes, can access these. And looking at what I googled, they are also used that way.

Maybe the input/output vs. internal really has to be understood as into/out of fastlane vs. inside of fastlane (including the `Fastfile`, plugins, lanes etc.). Then `lane_context` is fine to use for _any_ ruby code that is triggered by fastlane. (vs. in/out of the action that uses the lane_context).

> ENV variables would only play a role of public API - it is just raw idea 🤔

See above.

> Absolutely 💪 You think that would be possible to do by typing a command per lane, i.e. adding `SharedValues` summary table to `fastlane action [action_name]`? If so, could you please share the technical details? 🙏

Right now we have a `SharedValues` module in each action that should include all the 'keys'  set by this action. That sure is accessible somehow when creating the documentation and when `fastlane action ...` is run. How in detail, no idea 🤷‍♀️

**Comment by janpio** [comment]
> Not sure about ENV, but `debug` statement after any action will print all SharedValues currently available, not just from the last action (but of course it updates after every action)

@hydraSlav Can you share an example of what you mean? Maybe we are overlooking something here.

**Comment by hydraSlav** [comment]
> @hydraSlav Can you share an example of what you mean? Maybe we are overlooking something here.

This file

```
debug
match
debug
```

Will produce

```
[09:57:55]: -------------------
[09:57:55]: --- Step: debug ---
[09:57:55]: -------------------
Lane Context
{:DEFAULT_PLATFORM=>:ios, :PLATFORM_NAME=>:ios, :LANE_NAME=>"ios test5"}
[09:57:55]: -------------------
[09:57:55]: --- Step: match ---
[09:57:55]: -------------------
[...]
[MATCH OUTPUT HERE]
[...]
INFO [2019-02-25 09:58:07.30]: -------------------
INFO [2019-02-25 09:58:07.30]: --- Step: debug ---
INFO [2019-02-25 09:58:07.30]: -------------------
Lane Context
{:DEFAULT_PLATFORM=>:ios, :PLATFORM_NAME=>:ios, :LANE_NAME=>"ios test5", :SIGH_PROFILE_TYPE=>"development", :MATCH_PROVISIONING_PROFILE_MAPPING=>{"com.blah"=>"match Development com.blah"}}
```

What I was saying is the `debug` statement prints out all available lane variables at that particular time (across all actions/lanes)

Maybe I didn't understand the original issue by @mollyIV

**Comment by janpio** [comment]
TIL https://docs.fastlane.tools/actions/debug/ 

So using `debug` can definitely be a part of the solution to the general problem - it shows the content of lane_context during usage 👍

**Comment by mollyIV** [comment]
👋 @janpio @hydraSlav, a few more findings:

```ruby
bundle exec fastlane action download_dsyms
```

✨prints an interesting output among the other tables:
![image](https://user-images.githubusercontent.com/10795657/54178289-8f822e00-4495-11e9-8c77-085bdc055a8d.png)

### Implementation details

`download_dsyms` declares `SharedValues`
https://github.com/fastlane/fastlane/blob/66ca95528c264b2821e1a7caa78979677f561a2c/fastlane/lib/fastlane/actions/download_dsyms.rb#L3-L5

It sets the declared consts in a `lane_context`:
https://github.com/fastlane/fastlane/blob/66ca95528c264b2821e1a7caa78979677f561a2c/fastlane/lib/fastlane/actions/download_dsyms.rb#L118-L119

And last but not least please have a look at `output` declaration:

https://github.com/fastlane/fastlane/blob/66ca95528c264b2821e1a7caa78979677f561a2c/fastlane/lib/fastlane/actions/download_dsyms.rb#L235-L239

### Conclusion

Some actions (dunno how many 🤔) are missing `SharedValues` declaration in their `output` method even though they set them. I think that could be a good starting point - we could start with i.e. `match`  action and fill the missing `SharedValues` in `output` method WDYT? 

Still this design is error prone - it's very easy to miss adding `SharedValues` to `output` method. We should think about how to do it better by 🤖Maybe [rubocop](https://github.com/rubocop-hq/rubocop)?

**Comment by joshdholtz** [comment]
This issue totally slipped through my inbox and I have thoughts here 😱 There have a been a lot of thoughts here so I will just give my important parts.

## ENV (Environment variables)
In my opinion, `ENV` should not be used as a memory store to store anything during runtime and use an interprocess communication tool. `ENV` should also be used to set values that tools need to run (like using `.env` files to configure your fastlane execution environment and to configure actions/tools).

`ENV` variables can contain secret information (like API keys and passwords) so we also probably don't want to have an way to print all of these out (or if we do we will need to find a way to hide the secret ones)

## SharedValues / lane_context
This is what should be used to as a memory store to pass information back after a process executes and between other actions. The big benefit to putting this here is there is a separation of concerns between "input" from `ENV` and "output" from `SharedValues`/`lane_context`. 

The `SharedValues`/`lane_context` also gets the benefit of having to store objects (like arrays and hashes) while `ENV` can only store primitive data types otherwise it will 💥 

## TL;DR
`ENV` - input
`SharedValues`/`lane_context` - output

**Comment by janpio** [comment]
@joshdholtz What you write makes total sense. But does it match what fastlane does _right now_ in any way? I unfortunately don't think so. 

Some actions also seem to "export" parameters into ENV variables for... no idea :/ (Usually via a block that was thought to verify the parameter: https://github.com/fastlane/fastlane/blob/1af33c0884e783867bdc07a324280ec6bef124ad/produce/lib/produce/options.rb#L126-L128 or more: https://github.com/fastlane/fastlane/search?q=%22verify_block%3A+proc+do+|value|+ENV%22)

**Comment by janpio** [comment]
@mollyIV Regarding https://github.com/fastlane/fastlane/issues/14255#issuecomment-471872622

One thing that could probably be programmatically checked are differences between `output` and `module SharedValues`.

Checking for usage of `Actions.lane_context` in code and if the required `SharedValues` or `output` entries are there might be trickier, because often the actual code is not in the action itself - but this might still be worth an investigation (and maybe a "requirement"/force the developer to define it different if it is?)

Is `output` actually used consistently in many actions? If not, a rubocop rule would be the _perfect_ way to enforce that. And you can actually get pretty creative in rubocop, @revolter and @milch did awesome stuff in the past here.

**Comment by janpio** [comment]
Nah, we want to do something or multiple things about this @fastlane-bot

**Comment by mollyIV** [comment]
Alright, we merged https://github.com/fastlane/fastlane/pull/14610 💪 Let's figure out what can we do next to move this ticket forward, any ideas? 🤔 

Maybe there's something we can do with the input: `ENV - input`?

**Comment by janpio** [comment]
We could try to output the output in a table in the docs...

**Comment by mollyIV** [comment]
The issue has been created: [`The `output` table is visible in the action's documentation`](https://github.com/fastlane/fastlane/issues/14730) 💪

**Comment by janpio** [comment]
Possibly related thing I just noticed when debugging another issue:

https://github.com/fastlane/fastlane/blob/457c5d647c77f0e078dafa5129da616914e002c5/fastlane/lib/fastlane/actions/sync_code_signing.rb#L1-L38

This code defines `SharedValues::MATCH_PROVISIONING_PROFILE_MAPPING` and uses it, but also does `Actions.lane_context[SharedValues::SIGH_PROFILE_TYPE] = profile_type` - which would currently not be documented in any (automated) way I think. Not sure if this is common, but thought might be itneresting in this context.

**Comment by mollyIV** [comment]
Nice catch @janpio 💪 

I think one solution might be to improve the rubocop rule we created for `SharedValues`: 

check if there's `Actions.lane_context[SharedValues::CONST_NAME] = ...` assignment in a code but it is not declared in `module SharedValues`. 

WDYT? 🤔

**Comment by janpio** [comment]
That sounds awesome!

Although: Is it valid that `module SharedValues` lists a value that is normally created somewhere else? Can the same "value" be defined in two places? I assumed this had not only documentation value, but really defined some code.

**Comment by mollyIV** [comment]
👋 it's been a while since the last activity in this thread 😉 

A new PR has been created: https://github.com/fastlane/fastlane/pull/14892 💪

**Comment by janpio** [comment]
Pss bot, we are still working on this.

**Comment by mollyIV** [comment]
Still working on that @fastlane-bot ✨

**Comment by mollyIV** [comment]
ping @fastlane-bot

**Comment by joshdholtz** [comment]
Opening again cause this is important 😊

**Comment by mollyIV** [comment]
ping @fastlane-bot
